src/rgw/rgw_sync_module_pubsub.cc

```diff
@@ -0,0 +1,1443 @@
+          ldout(sync_env->cct, 20) << "pubsub: bucket create: using user info: " << json_str("obj", *sub->env->data_user_info, true) << dendl;
+          yield call(new RGWBucketCreateLocalCR(sync_env->async_rados,
+                                                sync_env->store,
+                                                create_bucket));
+          if (retcode < 0) {
+            ldout(sync_env->cct, 0) << "ERROR: failed to create bucket: " << "tenant="
+              << get_bucket_info.tenant << " name=" << get_bucket_info.bucket_name << ": ret=" << retcode << dendl;
+            return set_cr_error(retcode);
+          }
+
+          /* second iteration: we got -ENOENT and created a bucket */
+        }
+
+        /* failed twice on -ENOENT, unexpected */
+        ldout(sync_env->cct, 0) << "ERROR: failed to create bucket " << "tenant=" << get_bucket_info.tenant
+          << " name=" << get_bucket_info.bucket_name << dendl;
+        return set_cr_error(-EIO);
+      }
+      return 0;
+    }
+  };
+
+  using PushCR = RGWPostRESTResourceCR<rgw_pubsub_event, int>;
+
+  class StoreEventCR : public RGWCoroutine {
+    RGWDataSyncEnv *sync_env;
+    PSSubscriptionRef sub;
+    EventRef event;
+    PSEvent pse;
+    PSConfigRef& conf;
+    PSSubConfigRef& sub_conf;
+    rgw_object_simple_put_params put_obj;
+    string oid_prefix;
+    int i;
+  public:
+    StoreEventCR(RGWDataSyncEnv *_sync_env,
+                 PSSubscriptionRef& _sub,
+                 EventRef& _event) : RGWCoroutine(_sync_env->cct),
+                                     sync_env(_sync_env),
+                                     sub(_sub),
+                                     event(_event),
+                                     pse(_event),
+                                     conf(sub->env->conf),
+                                     sub_conf(sub->sub_conf) {
+      oid_prefix = sub->sub_conf->data_oid_prefix;
+    }
+
+    int operate() override {
+      reenter(this) {
+
+        put_obj.bucket = sub->bucket;
+        put_obj.key = rgw_obj_key(oid_prefix + pse.id());
+
+        pse.format(&put_obj.data);
+       
+        {
+          bufferlist bl;
+          pse.encode_event(bl);
+          bufferlist bl64;
+          bl.encode_base64(bl64);
+          put_obj.user_data = bl64.to_str();
+        }
+        
+        yield call(new RGWObjectSimplePutCR(sync_env->async_rados,
+                                            sync_env->store,
+                                            put_obj));
+        if (retcode < 0) {
+          ldout(sync_env->cct, 0) << "ERROR: failed to store event: " << put_obj.bucket << "/" << put_obj.key << " ret=" << retcode << dendl;
+          return set_cr_error(retcode);
+        }
+
+        if (sub->push.conn) {
+          yield {
+            rgw_http_param_pair params[] = {
+              { nullptr, nullptr }
+            };
+
+            call(new PushCR(sync_env->cct, sub->push.conn.get(),
```

Comment:
do you think it's worth trying to run this and the RGWObjectSimplePutCR in parallel?

Reply:
Not really. Push is not recommended.